core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -637,11 +639,12 @@ private[log] class Cleaner(val id: Int,
    * @param segment The segment to index
    * @param map The map in which to store the key=>offset mapping
    *
-   * @return The final offset covered by the map
+   * @return The final offset covered by the map or -1 if the map is full
    */
   private def buildOffsetMapForSegment(topicAndPartition: TopicAndPartition, segment: LogSegment, map: OffsetMap): Long = {
     var position = 0
     var offset = segment.baseOffset
+    val maxDesiredMapSize = (map.slots * this.dupBufferLoadFactor).toInt
```

Comment:
Is ```dupBufferLoadFactor``` useful after this change? You could modify ```OffsetMap.put``` to return ```Boolean``` to indicate if the put worked or map was full.